### ISSUE TITLE
GPG verification of packages

### ISSUE THREAD

**Issue by richardotis** [conda-forge/conda-smithy#254]
PyPI supports uploading GPG signatures along with packages. Example: https://pypi.io/packages/source/p/pycalphad/pycalphad-0.4.1.tar.gz.asc

It would be nice to be able to lock down a feedstock so that it will only allow rendering packages signed with a specified key.

(When you start thinking about how package updating would work if done properly, it seems to get complicated pretty fast: https://theupdateframework.github.io/ -- what I'm talking about here is only about securing the link between PyPI and anaconda.org via conda-forge, without concerning ourselves with the link to the end user.)

**Comment by jakirkham** [comment]
It is totally fine to have this issue here as this is where this change would need to happen. Though there is some discussion in issue ( https://github.com/conda-forge/conda-forge.github.io/issues/129 ) about this same idea.

**Comment by jakirkham** [comment]
Again no worries. Happy to have it here. 😄

**Comment by richardotis** [comment]
My first thought is that it could be sufficient to keep the relevant public key in the feedstock repo, plus an extra step to download and verify the signature of the package. On the other hand, it seems like the public key ought to be kept outside the feedstock to prevent the signature verification from being trivially defeated by the recipe maintainer.

Perhaps there should be a separate repo, `conda-forge/authorized-keys`, with restricted push access. Have some out-of-band process for verifying PRs to that repo, and keep public keys in a tree structure like `packagename/release-key-prod.pub`.

During feedstock rendering, conda-smithy would clone `authorized-keys` and check for the presence of any keyfiles in the directory matching the specified package. If there are any keyfiles, download the signature file indicated in the `signature` field of the feedstock (alternatively we could store the latest signature file in the feedstock repo), and verify the package file against the signature. Additionally, verify that the signature was made with one of the keys in the corresponding package directory from `authorized-keys``.

**Comment by jakirkham** [comment]
This is an old one, but may still be relevant.

cc @beckermr (who has been doing some related work)

**Comment by beckermr** [comment]
This is a feature that has to be implemented on the anaconda side. 

We have implemented an alternative procedure that can be managed on our side. It will be rolled out in a month or so.

**Comment by jakirkham** [comment]
So wdyt close in a month after our alternative is complete? Leave open until Anaconda implements it and then we implement support?

**Comment by beckermr** [comment]
Let's leave it open until/if we see something from anaconda.

**Comment by wolfv** [comment]
we're also pretty interested in implementing this in mamba and some alternative server :)
would be good if we get `gpgme` to work properly on OS X and Windows for that.

We already have the package on conda-forge: https://github.com/conda-forge/gpgme-feedstock/

**Comment by beckermr** [comment]
Cool! Note that package signing only solves part of the issue here. We would still need a way to scope public keys to specific outputs on anaconda.org in order to use package signing to validate that an output was produced by us and that it is ok for a given feedstock. Right now we use an API key that we issue for this.

**Comment by wolfv** [comment]
yep, sure -- package signing would have to be implemented as part of the conda server.
